tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/MsTestV2/UnitTestRunnerRunSingleTestAsyncIntegration3_8.cs

```diff
@@ -0,0 +1,141 @@
+// </copyright>
+#nullable enable
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+using Datadog.Trace.Ci;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.DuckTyping;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.MsTestV2;
+
+/// <summary>
+/// Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.UnitTestRunner.RunSingleTest calltarget instrumentation
+/// </summary>
+[InstrumentMethod(
+    AssemblyName = "Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter",
+    TypeName = "Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.UnitTestRunner",
+    MethodName = "RunSingleTestAsync",
+    ReturnTypeName = "System.Threading.Tasks.Task`1[Microsoft.VisualStudio.TestTools.UnitTesting.TestResult[]]",
+    ParameterTypeNames = ["Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel.TestMethod", "System.Collections.Generic.IDictionary`2[System.String,System.Object]", "Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging.IMessageLogger"],
+    MinimumVersion = "14.0.0",
+    MaximumVersion = "14.*.*",
+    IntegrationName = MsTestIntegration.IntegrationName)]
+[Browsable(false)]
+[EditorBrowsable(EditorBrowsableState.Never)]
+// ReSharper disable once InconsistentNaming
+public static class UnitTestRunnerRunSingleTestAsyncIntegration3_8
+{
+    internal static CallTargetState OnMethodBegin<TTarget, TTestMethod, TMessageLogger>(TTarget instance, TTestMethod testMethod, IDictionary<string, object?> testContextProperties, TMessageLogger messageLogger)
+    {
+        return new CallTargetState(null, new StrongBox<MethodInfoCacheItem?>(MsTestIntegration.IsTestMethodRunnableThreadLocal.Value));
+    }
+
+    internal static CallTargetReturn<TReturn?> OnMethodEnd<TTarget, TReturn>(TTarget instance, TReturn? returnValue, Exception? exception, in CallTargetState state)
+    {
+        if (state.State is StrongBox<MethodInfoCacheItem?> box)
+        {
+            box.Value = MsTestIntegration.IsTestMethodRunnableThreadLocal.Value;
+        }
+
+        MsTestIntegration.IsTestMethodRunnableThreadLocal.Value = null;
+        return new CallTargetReturn<TReturn?>(returnValue);
+    }
+
+    internal static TReturn? OnAsyncMethodEnd<TTarget, TReturn>(TTarget instance, TReturn? returnValue, Exception? exception, in CallTargetState state)
+    {
+        if (instance is null || !MsTestIntegration.IsEnabled)
+        {
+            return returnValue;
+        }
+
+        MethodInfoCacheItem? methodInfoCacheItem;
+        if (state.State is StrongBox<MethodInfoCacheItem?> box)
+        {
+            methodInfoCacheItem = box.Value ?? MsTestIntegration.IsTestMethodRunnableThreadLocal.Value;
+        }
+        else
+        {
+            methodInfoCacheItem = MsTestIntegration.IsTestMethodRunnableThreadLocal.Value;
+        }
```

Comment:
Maybe this is a little neater? Undecided
```suggestion
        if (state.State is StrongBox<MethodInfoCacheItem?> box && box.Value is { } value)
        {
            methodInfoCacheItem = value;
        }
        else
        {
            methodInfoCacheItem = MsTestIntegration.IsTestMethodRunnableThreadLocal.Value;
        }
```

Reply:
I like that, changed in https://github.com/DataDog/dd-trace-dotnet/pull/6686/commits/a565f8452c22f2fabbeaf9e33fad85bd090faa55